### DISCUSSION TITLE
[RFC] Routes Manifest for Type Safety and to Decouple Route Names from the URL

### DISCUSSION THREAD

**Discussion by flybayer** [vercel/next.js#32769] [Ideas]
# Problem

There are currently three key issues with the Next.js workflow of using string literals for Link href's.

1. Too easy to cause broken links, and no warning or error when it happens
2. When moving a page to different a URL path, you have to go manually update every instance of a link to that page.
2. No type safety or autocomplete

# Solution

Add a Prisma-style code generation feature that writes a fully typed routes manifest file inside node_modules.

1. Enables full type safety for routing, including autocomplete for JS users. Requires you to pass all dynamic parameters for dynamic routes.
2. Prevents unknown broken links
3. Allows you to move a page to different URL without updating any links to the page

```ts
// Assume you have a page at pages/products/[productId].tsx
export default function ProductsPage() { ...

// You can then use Routes...
import Routes from "next/routes"

// ...to refer to it by name (Note: `productId` is a required parameter since it's in the route path)
<Link href={Routes.ProductsPage({ productId: 123 })} />

// ...instead of hard coding a string!
<Link href={`/products/${123}`} />
```

Query parameters can be specified together with route parameters.

```ts
// instead of ...
<Link href={`/products/${pid}?offerCode=capybara`} />

// ... you can do:
<Link href={Routes.Product({ pid, offerCode: "capybara" })} />
```

- All places Next.js accepts an `href` will also work with the routes manifest like `getServerSideProps` redirect.
- Fully optional feature
- Fully backwards compatible

# Context

[Blitz.js is pivoting to a framework agnostic toolkit](https://github.com/blitz-js/blitz/discussions/3075). **This is a highly loved Blitz feature that we strongly feel will benefit all Next.js users.** While we can keep it in the new Blitz toolkit, the DX will be degraded without being hooked into the Next.js compiler for snappy, seamless updates in development.

**Comment by matamatanot** [comment]
For your reference, [pathpida](https://github.com/aspida/pathpida).

**Comment by kylemh** [comment]
Part of the creation story with Routes could also include the generating of `robots.txt` + `sitemap.xml` too 👀

**Comment by ckastbjerg** [comment]
> 💁 Apologies for the self promotion

I authored a library specifically to address these challenges: https://github.com/ckastbjerg/next-type-safe-routes. I would much rather have Next.js support this out of the box, but just in case anyone is looking for a solution here and now. And maybe some of the ideas of the library could also be used as input for this discussion :)

**Comment by flybayer** [comment]
Here's how the current Blitz implementation works.

![image](https://user-images.githubusercontent.com/8813276/147286402-c95a182e-8df5-48d8-90f9-197886166c76.png)

**Reply by tawaliou** [reply]
Very interesting how you solve this problem.

Blitz generate this file from our Link for us ?

**Reply by timneutkens** [reply]
Based on the implementation shown above the immediate downside is that the generated file will grow indefinitely as it is not tree-shaken in any way, and even if it is tree-shaken to only routes used it'll still be less than ideal because it'd mean that the full file with routes that have been used would still be loaded on each page that uses one of the routes. One way to solve that is inlining them using a transform. However, that likely brings some complications with testings. Maybe `next/jest` will help solve that though. Definitely something to think about more for what the implementation will look like.

Having a typed way to generate urls would be useful, could see that being used for vercel.com where we have a lot of `next/link` usage. An example why the manifest is not ideal as-is is that on vercel.com we have 289 pages, if all of them were called `Home` with just `/` as the pathname that would already be 25KB minified before gzip, obviously there's a bunch of longer names than `Home` and deep nesting of the directory tree that would hold much longer pathnames

Overall I'm personally not opposed to this change if there's no performance impact, will have to check with the rest of the team though.

**Reply by timneutkens** [reply]
Thought about this a bit more and it seems introducing this would also cause a chicken/egg problem with TypeScript as you would no longer be able to run the type checking outside of Next.js as a step before building. Running `tsc` yourself would no longer be possible as the types have to be generated before type checking happens.

Just flagging it as I thought about it this morning.

**Reply by flybayer** [reply]
@timneutkens 
> One way to solve that is inlining them using a transform. However, that likely brings some complications with testings.

If in addition to the transform we generate the monolithic file, then the monolithic file can be used normally with tests, etc.

>  it seems introducing this would also cause a chicken/egg problem with TypeScript as you would no longer be able to run the type checking outside of Next.js as a step before building

For blitz we added a `blitz codegen` command for use cases where you want to just generate the types.

**Comment by kvnwolf** [comment]
I created a small utility some months ago that I use on my projects but never open sourced, in case you want to take ideas from it or someone wants to open source it.

Example tests/usage:

```ts
// index.test.ts
import { createRoute } from "."

const routes = {
  home: createRoute("/"),
  faq: createRoute("faq"),
  about: createRoute("about/"),
  termsAndConditions: createRoute("terms//and///conditions"),
  project: createRoute("projects/[p]"),
  task: createRoute("projects/[p]/tasks/[t]"),
  docs: createRoute("docs/[...p]"),
  files: createRoute("files/[[...p]]"),
  projects: createRoute("projects", {
    new: createRoute("new"),
    project: createRoute("[p]", {
      edit: createRoute("edit"),
      tasks: createRoute("tasks", {
        task: createRoute("[t]"),
      }),
    }),
  }),
}

test("sanitizes the route path", () => {
  expect(routes.home()).toBe("/")
  expect(routes.faq()).toBe("/faq")
  expect(routes.about()).toBe("/about")
  expect(routes.termsAndConditions()).toBe("/terms/and/conditions")
})

test("builds a route with dynamic params", () => {
  expect(routes.project({ p: "p1" })).toBe("/projects/p1")
  expect(routes.task({ p: "p1", t: "t1" })).toBe("/projects/p1/tasks/t1")
})

test("builds a route with catch-all params", () => {
  expect(routes.docs({ p: ["core"] })).toBe("/docs/core")
  expect(routes.docs({ p: ["core", "utils"] })).toBe("/docs/core/utils")
})

test("builds a route with optional catch-all params", () => {
  expect(routes.files({ p: [] })).toBe("/files")
  expect(routes.files({ p: ["images"] })).toBe("/files/images")
  expect(routes.files({ p: ["images", "rocket.png"] })).toBe("/files/images/rocket.png")
})

test("supports nested routes", () => {
  expect(routes.projects()).toBe("/projects")
  expect(routes.projects.new()).toBe("/projects/new")
  expect(routes.projects.project({ p: "p1" })).toBe("/projects/p1")
  expect(routes.projects.project.edit({ p: "p1" })).toBe("/projects/p1/edit")
  expect(routes.projects.project.tasks({ p: "p1" })).toBe("/projects/p1/tasks")
  expect(routes.projects.project.tasks.task({ p: "p1", t: "t1" })).toBe("/projects/p1/tasks/t1")
})
```

Implementation:
```ts
// index.ts
import { List, String, Union } from "ts-toolbelt"

export function createRoute<TRoutePath extends string, TSubRoutes extends SubRoutes | void = void>(
  routePath: TRoutePath,
  subRoutes?: TSubRoutes,
) {
  const routeBuilder = (params?: RouteParams) => replaceRouteParams(routePath, params)
  routeBuilder.__path__ = routePath

  if (subRoutes) {
    routeBuilder.__subRoutes__ = subRoutes

    for (const [subRouteName, subRouteBuilder] of Object.entries(subRoutes as SubRoutes)) {
      Object.assign(routeBuilder, {
        [subRouteName]: createRoute(
          `${routeBuilder.__path__}/${(subRouteBuilder as any).__path__}`,
          (subRouteBuilder as any).__subRoutes__,
        ),
      })
    }
  }

  return routeBuilder as any as RouteFactory<TRoutePath, TSubRoutes>
}

function replaceRouteParams(routePath: string, params?: RouteParams) {
  if (params) {
    const requiredParams =
      routePath.match(/\[{1,2}(\.{3})?[^\\[\]\\.]+\]{1,2}/g)?.map((value) => ({
        key: value.replace(/(\[|\]|\.)/g, ""),
        value,
      })) ?? []

    for (const { key, value } of requiredParams) {
      const paramValue = params[key]

      routePath = routePath.replace(
        value,
        paramValue instanceof Array ? paramValue.join("/") : paramValue,
      )
    }
  }

  return sanitizeRoutePath(routePath)
}

function sanitizeRoutePath(routePath: string) {
  if (!routePath.startsWith("/")) routePath = "/".concat(routePath)
  if (routePath.length > 1 && routePath.endsWith("/")) routePath = routePath.slice(0, -1)
  routePath = routePath.replace(/\/+/g, "/")
  return routePath
}

type RouteFactory<TRoutePath extends string, TSubRoutes extends SubRoutes | void = void> =
  RouteBuilder<TRoutePath> &
    (TSubRoutes extends SubRoutes ? ExpandSubRoutes<TRoutePath, TSubRoutes> : Record<never, never>)

type RouteBuilder<TRoutePath extends string> = (...args: RouteBuilderArgs<TRoutePath>) => string

type RouteBuilderArgs<
  TRoutePath extends string,
  TRouteParams extends RouteParams = InferRouteParamsFromPath<TRoutePath>,
> = Record<never, never> extends TRouteParams ? [] : [params: TRouteParams]

type RouteParams = Record<string | never, RouteParamType | never>
type RouteParamType = string | OneOrMoreStrings | ZeroOrMoreStrings
type OneOrMoreStrings = { 0: string } & string[]
type ZeroOrMoreStrings = string[]

type InferRouteParamsFromPath<
  TRoutePath extends string,
  TRoutePart extends string = List.UnionOf<String.Split<TRoutePath, "/">>,
  TRouteParam extends Record<string, RouteParamType> = InferRouteParamFromPart<TRoutePart>,
> = Union.Merge<TRouteParam>

type InferRouteParamFromPart<TRoutePart extends string> =
  TRoutePart extends `[[...${infer TRouteParamName}]]`
    ? { [key in TRouteParamName]: ZeroOrMoreStrings }
    : TRoutePart extends `[...${infer TRouteParamName}]`
    ? { [key in TRouteParamName]: OneOrMoreStrings }
    : TRoutePart extends `[${infer TRouteParamName}]`
    ? { [key in TRouteParamName]: string }
    : never

type SanitizeRoutePath<
  TRoutePath extends string,
  TRoutePart extends string = List.UnionOf<String.Split<TRoutePath, "/">>,
> = `/${String.Join<List.Filter<Union.ListOf<TRoutePart>, "">, "/">}`

type SubRoutes = { [name: string]: RouteFactory<any, any> }

type ExpandSubRoutes<TParentPath extends string, TSubRoutes extends SubRoutes> = {
  [TSubRoute in keyof TSubRoutes]: TSubRoutes[TSubRoute] extends RouteFactory<
    infer TSubRoutePath,
    infer TSubRouteSubRoutes
  >
    ? RouteFactory<SanitizeRoutePath<`${TParentPath}/${TSubRoutePath}`>, TSubRouteSubRoutes>
    : never
}

export type InferRouteParams<TRouteFactory extends RouteFactory<any>> =
  TRouteFactory extends RouteFactory<infer TRoutePath>
    ? InferRouteParamsFromPath<TRoutePath>
    : never
```

**Reply by tawaliou** [reply]
Look at Blitz or pathpida

**Comment by imcodingideas** [comment]
Thank you. Please get this into nextjs core ;)

**Comment by flybayer** [comment]
One other feature that's been requested for Blitz is ability to get the fully formed string.

With above blitz implementation, you get:
```ts
Routes.ProjectPage({projectId: 1}).pathname === '/projects/[projectId]'
```

But people want this too:

```ts
Routes.ProjectPage({projectId: 1}).href === '/projects/1'
```

**Reply by nandorojo** [reply]
@flybayer would it be hard to abstract the script Blitz uses into its own pathpida-like package for now?

**Reply by flybayer** [reply]
@nandorojo no, it wouldn't be hard. We may need to do that for blitz backward compatibility.

**Comment by vivek7405** [comment]
It would be great if we can have this feature in NextJs

**Reply by tawaliou** [reply]
Yes but I think Pathpida on their github said that they have a great support for Next.js

**Comment by tatethurston** [comment]
I looked into approaching this problem slightly differently and augment the types for `next/link` and `next/router` with an ambient type declaration. This avoids creating an additional runtime, which sidesteps the tree shaking issue that @timneutkens mentioned.

This is distributed on npm as [nextjs-routes](https://github.com/tatethurston/nextjs-routes).
